### PR TITLE
Labeling for anomalies #631

### LINKED ISSUE #631: Labeling for anomalies
It's for https://github.com/hastic/hastic-grafana-app/issues/281

* Endpoint to send labels
* Sending labeling to analytics (without algorithms update)

### PR DESCRIPTION
**Pull request by amper43**
fixes #631 

## Changes
- `seasonality` and `seasonalityPeriod` fields in analytic unit
- add / subtract segment data from each smoothed dataset period
- generate learning timerange for `Pattern` and `Anomaly` detectors using labeled OR deleted segments 

## Motivation
Imagine you created an anomaly analytic unit for a metric with seasonality where each period has a part outside the confidence interval:
image

Now you can set the seasonality period and label segment you don't want to be detected
Hastic-server will change the smoothed dataset (adding / subtracting segment data to each period) such that this "anomaly" will be inside the confidence interval
Example with 24 hours seasonality:
image

## Known issues
- Confidence interval is not shown correctly (https://github.com/hastic/hastic-grafana-app/issues/306)
- Duplication in `process_data` and `detect`
- Bounds for different segments are rewrited instead of adding

### CODE REVIEW

File: analytics/analytics/detectors/anomaly_detector.py [hunk 1]
```diff
@@ -22,16 +22,41 @@ def __init__(self, analytic_unit_id: AnalyticUnitId):
...
+        prepared_segments = []
+
+        new_cache = {
+            'confidence': payload['confidence'],
+            'alpha': payload['alpha']
+        }
+
+        if segments is not None:
+            seasonality = payload.get('seasonality')
+            assert seasonality is not None and seasonality > 0, f'{self.analytic_unit_id} got invalid seasonality {seasonality}'
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: let's split it into 2 lines

File: analytics/analytics/detectors/anomaly_detector.py [hunk 2]
```diff
@@ -22,16 +22,41 @@ def __init__(self, analytic_unit_id: AnalyticUnitId):
...
+
+            for segment in segments:
+                segment_len = (int(segment['to']) - int(segment['from']))
+                assert segment_len <= seasonality, f'seasonality {seasonality} must be great then segment length {segment_len}'
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: too long

File: analytics/analytics/detectors/anomaly_detector.py [hunk 3]
```diff
@@ -22,16 +22,41 @@ def __init__(self, analytic_unit_id: AnalyticUnitId):
...
+
+                from_index = utils.timestamp_to_index(dataframe, pd.to_datetime(segment['from'], unit='ms'))
+                to_index = utils.timestamp_to_index(dataframe, pd.to_datetime(segment['to'], unit='ms'))
+                segment_data = dataframe[from_index: to_index]
+                prepared_segments.append({'from': segment['from'], 'data': segment_data.value.tolist()})
+
+            time_step = utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][1]) - utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][0])
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: too long

File: analytics/analytics/detectors/anomaly_detector.py [hunk 4]
```diff
@@ -40,6 +65,25 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
         upper_bound = smoothed_data + cache['confidence']
         lower_bound = smoothed_data - cache['confidence']
 
+        if segments is not None:
+
+            seasonality = cache.get('seasonality')
+            assert seasonality is not None and seasonality > 0, f'{self.analytic_unit_id} got invalid seasonality {seasonality}'
+
+            data_start_time = utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][0])
+            time_step = utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][1]) - utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][0])
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: too long

File: analytics/analytics/detectors/anomaly_detector.py [hunk 5]
```diff
@@ -40,6 +65,25 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
...
+
+            for segment in segments:
+                seasonality_offset = (abs(segment['from'] - data_start_time) % seasonality) // time_step
+                seasonality_index = seasonality // time_step
+                #TODO: upper and lower bounds for segment_data
+                segment_data = utils.exponential_smoothing(pd.Series(segment['data']), 0.5)
+                upper_seasonality_curve = self.add_season_to_data(smoothed_data, segment_data, seasonality_offset, seasonality_index, True)
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: too long
also, for 2 lines above

File: server/src/controllers/analytics_controller.ts [hunk 6]
```diff
@@ -117,20 +117,20 @@ async function getQueryRange(
   analyticUnitId: AnalyticUnit.AnalyticUnitId,
   detectorType: AnalyticUnit.DetectorType
 ): Promise<TimeRange> {
-  if(detectorType === AnalyticUnit.DetectorType.PATTERN) {
+  if(
+    detectorType === AnalyticUnit.DetectorType.PATTERN ||
+    detectorType === AnalyticUnit.DetectorType.ANOMALY
+  ) {
     // TODO: find labeled OR deleted segments to generate timerange
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
```

File: analytics/analytics/utils/common.py [hunk 7]
```diff
@@ -351,6 +351,10 @@ def get_correlation(segments: list, av_model: list, data: pd.Series, window_size
             p_value_list.append(correlation[1])
     return correlation_list
 
+def get_lag_by_autocorrelation(data: pd.Series) -> int:
```
Comment (rozetko) [inline analytics/analytics/utils/common.py]: is it used anywhere?

File: analytics/analytics/detectors/anomaly_detector.py [hunk 8]
```diff
@@ -103,14 +151,43 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+    def add_season_to_data(self, data: pd.Series, segment: pd.Series, offset: int, seasonality: int, addition: bool) -> pd.Series:
+        #data - smoothed data to which seasonality will be added
+        #if additional == False -> segment is subtracted
+        len_smoothed_data = len(data)
+        for idx, _ in enumerate(data):
+            if idx - offset < 0:
+                continue
+            if (idx - offset) % seasonality == 0:
+                if addition:
+                    data = data.add(pd.Series(segment.values, index=segment.index + idx), fill_value=0)
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
                    data = data.add(pd.Series(segment.values, index = segment.index + idx), fill_value=0)
```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 9]
```diff
@@ -103,14 +151,43 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+                else:
+                    data = data.add(pd.Series(segment.values * -1, index=segment.index + idx), fill_value=0)
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
                    data = data.add(pd.Series(segment.values * -1, index = segment.index + idx), fill_value = 0)
```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 10]
```diff
@@ -103,14 +151,43 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
         smoothed_dataset = list(zip(timestamps, smoothed.values.tolist()))
         result = ProcessingResult(smoothed_dataset)
         return result
 
-    def merge_segments(self, segments: List[Segment]) -> List[Segment]:
-        segments = utils.merge_intersecting_segments(segments)
-        return segments
+    def add_season_to_data(self, data: pd.Series, segment: pd.Series, offset: int, seasonality: int, addition: bool) -> pd.Series:
+        #data - smoothed data to which seasonality will be added
+        #if additional == False -> segment is subtracted
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
        #if addition == True -> segment is added
        #if addition == False -> segment is subtracted
```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 11]
```diff
@@ -22,16 +22,41 @@ def __init__(self, analytic_unit_id: AnalyticUnitId):
...
+            seasonality = payload.get('seasonality')
+            assert seasonality is not None and seasonality > 0, f'{self.analytic_unit_id} got invalid seasonality {seasonality}'
+
+            for segment in segments:
+                segment_len = (int(segment['to']) - int(segment['from']))
+                assert segment_len <= seasonality, f'seasonality {seasonality} must be great then segment length {segment_len}'
+
+                from_index = utils.timestamp_to_index(dataframe, pd.to_datetime(segment['from'], unit='ms'))
+                to_index = utils.timestamp_to_index(dataframe, pd.to_datetime(segment['to'], unit='ms'))
+                segment_data = dataframe[from_index: to_index]
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]:
```suggestion
                segment_data = dataframe[from_index : to_index]
```

File: analytics/analytics/detectors/anomaly_detector.py [hunk 12]
```diff
@@ -40,6 +65,25 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
...
+            assert seasonality is not None and seasonality > 0, f'{self.analytic_unit_id} got invalid seasonality {seasonality}'
+
+            data_start_time = utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][0])
+            time_step = utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][1]) - utils.convert_pd_timestamp_to_ms(dataframe['timestamp'][0])
+
+            for segment in segments:
+                seasonality_offset = (abs(segment['from'] - data_start_time) % seasonality) // time_step
+                seasonality_index = seasonality // time_step
+                #TODO: upper and lower bounds for segment_data
+                segment_data = utils.exponential_smoothing(pd.Series(segment['data']), 0.5)
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: why 0.5?

File: analytics/analytics/detectors/anomaly_detector.py [hunk 13]
```diff
@@ -40,6 +65,25 @@ def detect(self, dataframe: pd.DataFrame, cache: Optional[ModelCache]) -> Detect
...
+                upper_seasonality_curve = self.add_season_to_data(smoothed_data, segment_data, seasonality_offset, seasonality_index, True)
+                lower_seasonality_curve = self.add_season_to_data(smoothed_data, segment_data, seasonality_offset, seasonality_index, False)
+                assert len(smoothed_data) == len(upper_seasonality_curve), f'len smoothed {len(smoothed_data)} != len seasonality {len(seasonality_curve)}'
+                upper_bound = upper_seasonality_curve + cache['confidence']
+                lower_bound = lower_seasonality_curve - cache['confidence']
+
```
Comment (rozetko) [inline analytics/analytics/detectors/anomaly_detector.py]: `process_data` is almost identical to `detect`
Let's move a common part to function

File: analytics/analytics/detectors/anomaly_detector.py [hunk 14]
```diff
@@ -22,16 +22,41 @@ def __init__(self, analytic_unit_id: AnalyticUnitId):
         self.bucket = DataBucket()
 
     def train(self, dataframe: pd.DataFrame, payload: Union[list, dict], cache: Optional[ModelCache]) -> ModelCache:
+        segments = payload.get('segments')
+        prepared_segments = []
```
Comment (amper43) [inline analytics/analytics/detectors/anomaly_detector.py]: move it into `if` statement

File: analytics/analytics/detectors/anomaly_detector.py [hunk 15]
```diff
@@ -103,14 +151,44 @@ def concat_detection_results(self, detections: List[DetectionResult]) -> Detecti
...
+
         # TODO: exponential_smoothing should return dataframe with related timestamps
-        smoothed = utils.exponential_smoothing(data['value'], cache['alpha'], cache.get('lastValue'))
-        timestamps = utils.convert_series_to_timestamp_list(data.timestamp)
+        smoothed = utils.exponential_smoothing(dataframe['value'], cache['alpha'], cache.get('lastValue'))
+
+        if segments is not None:
+
+            seasonality = cache.get('seasonality')
+            assert seasonality is not None and seasonality > 0, f'{self.analytic_unit_id} got invalid seasonality {seasonality}'
```
Comment (amper43) [inline analytics/analytics/detectors/anomaly_detector.py]: too long

File: analytics/analytics/detectors/anomaly_detector.py (hunk 12 above)
Comment (VargBurz) [reply analytics/analytics/detectors/anomaly_detector.py]: basic alpha for smoothing

**Pull request merged**